packages/react-reconciler/src/ReactFiberRootScheduler.js

```diff
@@ -0,0 +1,465 @@
+        break;
+      default:
+        schedulerPriorityLevel = NormalSchedulerPriority;
+        break;
+    }
+
+    const newCallbackNode = scheduleCallback(
+      schedulerPriorityLevel,
+      performConcurrentWorkOnRoot.bind(null, root),
+    );
+
+    root.callbackPriority = newCallbackPriority;
+    root.callbackNode = newCallbackNode;
+    return newCallbackPriority;
+  }
+}
+
+export type RenderTaskFn = (didTimeout: boolean) => RenderTaskFn | null;
+
+export function getContinuationForRoot(
+  root: FiberRoot,
+  originalCallbackNode: mixed,
+): RenderTaskFn | null {
+  // This is called at the end of `performConcurrentWorkOnRoot` to determine
+  // if we need to schedule a continuation task.
+  //
+  // Usually `scheduleTaskForRootDuringMicrotask` only runs inside a microtask;
+  // however, since most of the logic for determining if we need a continuation
+  // versus a new task is the same, we cheat a bit and call it here. This is
+  // only safe to do because we know we're at the end of the browser task.
+  // So although it's not an actual microtask, it might as well be.
+  scheduleTaskForRootDuringMicrotask(root, now());
+  if (root.callbackNode === originalCallbackNode) {
+    // The task node scheduled for this root is the same one that's
+    // currently executed. Need to return a continuation.
+    return performConcurrentWorkOnRoot.bind(null, root);
+  }
+  return null;
+}
+
+const fakeActCallbackNode = {};
+
+function scheduleCallback(
+  priorityLevel: PriorityLevel,
+  callback: RenderTaskFn,
+) {
+  if (__DEV__ && ReactCurrentActQueue.current !== null) {
+    // Special case: We're inside an `act` scope (a testing utility).
+    // Instead of scheduling work in the host environment, add it to a
+    // fake internal queue that's managed by the `act` implementation.
+    ReactCurrentActQueue.current.push(callback);
+    return fakeActCallbackNode;
+  } else {
+    return Scheduler_scheduleCallback(priorityLevel, callback);
+  }
+}
+
+function cancelCallback(callbackNode: mixed) {
+  if (__DEV__ && callbackNode === fakeActCallbackNode) {
+    // Special `act` case: check if this is the fake callback node used by
+    // the `act` implementation.
+  } else if (callbackNode !== null) {
+    Scheduler_cancelCallback(callbackNode);
+  }
+}
+
+function scheduleImmediateTask(cb: () => mixed) {
+  if (__DEV__ && ReactCurrentActQueue.current !== null) {
+    // Special case: Inside an `act` scope, we push microtasks to the fake `act`
+    // callback queue. This is because we currently support calling `act`
+    // without awaiting the result. The plan is to deprecate that, and require
+    // that you always await the result so that the microtasks have a chance to
+    // run. But it hasn't happened yet.
+    ReactCurrentActQueue.current.push(() => {
+      cb();
+      return null;
+    });
+  }
+
+  // TODO: Can we land supportsMicrotasks? Which environments don't support it?
```

Comment:
I don't think so, React Native does not support microtasks. Hermes is working on it, but it doesn't exist in JSC. If we moved this to the host config, then RN could probably do a user-land fallback for JSC if microtasks were not available. I think I originally had it layered so it was all in the host config so +1.

Previous discussion here: https://github.com/facebook/react/pull/20658

Reply:
I think we could argue that Immediate tasks should run _before_ microtasks, which is essentially the only way we use the immediate priority now, so you could argue this belongs in the Scheduler :)

Reply:
Ok I'll just leave this as-is for now then